engine/client/secretprovider/secretprovider.go

```diff
@@ -4,34 +4,67 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"net/url"
 	"strings"
+	"time"
 
 	"github.com/moby/buildkit/session/secrets"
 	"google.golang.org/grpc"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 )
 
+type dataWithTTL struct {
+	expiresAt time.Time
+	data      map[string]any
+}
+
 type SecretResolver func(context.Context, string) ([]byte, error)
 
 var resolvers = map[string]SecretResolver{
 	"env":   envProvider,
 	"file":  fileProvider,
 	"cmd":   cmdProvider,
 	"op":    opProvider,
-	"vault": vaultProvider,
+	"vault": getVaultProvider(0), // no ttl for backward compatibility
 }
 
-func ResolverForID(id string) (SecretResolver, string, error) {
-	scheme, path, ok := strings.Cut(id, "://")
-	if !ok {
-		return nil, "", fmt.Errorf("parse %q: malformed id", id)
+func getSecretResolverWithTTL(scheme string, ttl time.Duration) (SecretResolver, error) {
+	if scheme == "vault" && ttl > 0 {
```

Comment:
We shouldn't have to special case `vault` in `secretprovider.go`. All the vault logic should go in `vault.go`.